### PR TITLE
Scalar multiplication

### LINKED ISSUE #348: [Proposal] Scalar Multiplication * for FactoredMatrix
Is there any particular reason why multiplication with a scalar (*) should not be implemented for FactoredMatrix?
I need it in my research, and while there are easy workarounds I would like to add it to streamline working with FactoredMatrix.
In the absence of counterarguments I would implement it and the according tests myself and submit a PR.

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/neelnanda-io/transformerlens/issues) in the repo (**required**)

### PR DESCRIPTION
**Pull request by matthiasdellago**
# Description

Implemented __mul__ and __rmul__ for Factoredmatrix to support scalar multiplication with *.
Supports floats, ints and single-element Tensors.
Approriate error messages and test are in place.

Fixes #348 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

**Comment by neelnanda-io** [comment]
Thanks! It feels a bit inelegant to arbitrarily multiply the matrix on the right (rather than the left, or sqrt(scalar) on both), but it's probably best as is - users aren't supposed to use individual components of the factoredmatrix in general, and this avoids numerical errors of sqrt. 

I left some nits, but overall LGTM

**Comment by neelnanda-io** [comment]
Looks like you have some formatting changes

File: tests/unit/factored_matrix/test_multiply_by_scalar.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+# This test function is parametrized with different types of scalars, including non-scalar tensors and arrays, to check that the correct errors are raised.
+# Considers cases with and without leading dimensions as well as left and right multiplication.
+@pytest.mark.parametrize(
+    "scalar, error_expected", 
+    [
+        # Test cases with different types of scalar values.
+        (torch.rand(1), None),  # 1-element Tensor. No error expected.
+        (float(torch.rand(1).item()), None),  # float. No error expected.
+        (int(torch.randint(1, 10, (1,)).item()), None),  # int. No error expected.
```
[the remark this answers is not in the document]
Comment (matthiasdellago) [reply tests/unit/factored_matrix/test_multiply_by_scalar.py:16]: You're right, I just wanted to minimise imports. I'll take care of it.

File: tests/unit/factored_matrix/test_multiply_by_scalar.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        with pytest.raises(error_expected):
+            if multiply_from_left:
+                _ = fm * scalar
+            else:
+                _ = scalar * fm
+    else:
+        # If no error is expected, check that the multiplication results in the correct value.
+        # Use FactoredMatrix.AB to calculate the product of the two factor matrices before comparing with the expected value.
+        if multiply_from_left:
+            assert_close((fm * scalar).AB, (a @ b) * scalar)
```
[the remark this answers is not in the document]
Comment (matthiasdellago) [reply tests/unit/factored_matrix/test_multiply_by_scalar.py:60]: Sure, I think that keeping the test agnostic to how exactly it is implemented is prettier, but if you prefer to have a test for (fm * scalar).A I will add it.

**Comment by matthiasdellago** [comment]
> Thanks! It feels a bit inelegant to arbitrarily multiply the matrix on the right (rather than the left, or sqrt(scalar) on both), but it's probably best as is - users aren't supposed to use individual components of the factoredmatrix in general, and this avoids numerical errors of sqrt.

Agreed. Aside from the numerical error my main resoning was to reduce the amount of computation required.

> Looks like you have some formatting errors

Yes, I forgot to set up my development environment properly. I'm on it.

**Comment by matthiasdellago** [comment]
Resolved the nits, and fixed the formatting. :+1: 
Thanks for making TransformerLens btw, i really enjoy using it! :)

**Comment by jbloomAus** [comment]
Thanks!

**Pull request merged**